staging/src/k8s.io/apiserver/pkg/server/options/audit.go

```diff
@@ -436,13 +502,17 @@ func (o *AuditLogOptions) getWriter() io.Writer {
 	return w
 }
 
-func (o *AuditLogOptions) applyTo(c *server.Config) error {
+func (o *AuditLogOptions) applyTo(c server.Config) audit.Backend {
 	if w := o.getWriter(); w != nil {
+		if c.AuditPolicyChecker == nil {
+			klog.V(2).Info("No audit policy file provided, no events will be recorded for log backend")
+			return nil
+		}
 		groupVersion, _ := schema.ParseGroupVersion(o.GroupVersionString)
 		log := pluginlog.NewBackend(w, o.Format, groupVersion)
 		log = o.BatchOptions.wrapBackend(log)
 		log = o.TruncateOptions.wrapBackend(log, groupVersion)
-		c.AuditBackend = appendBackend(c.AuditBackend, log)
+		return appendBackend(c.AuditBackend, log)
```

Comment:
I'm finding the changes to the applyTo methods hard to follow:
* AuditOptions#applyTo does not modify config, and returns a policy checker
* AuditLogOptions#applyTo does not modify the config, but returns a union backend, but only if a checker is set
* AuditWebhookOptions#applyTo does not modify config, and returns a non-unioned webhook-only backend, but only if a checker is set
* AuditDynamicOptions#applyTo does not modify the config, and returns both a backend (non-unioned) and a checker

Reply:
before this PR, the individual audit applyTo methods were independent, and could be called in any order, and chained their own contribution into the server config

after this PR, they must be called in a specific order, and no longer compose audit options into the config, but return things that are then combined in AuditOptions#ApplyTo?

I'm also finding the `if o.DynamicOptions.enabled()` block a bit impenetrable.

Reply:
@deads2k often has good feedback about structuring options -> application -> config

Reply:
so one of the tricky pieces here was that the dynamic backend needs to wrap the log and webhook backends if they exist, it also needs to be wrapped in the truncate backend if it exists. This made order relevant, I definitely preferred the independent nature of these before, but its not possible with the current reqs. I can fix up some of the naming and add better commenting

Reply:
I think I agree. Having the individual methods just return building blocks and assembling them all in AuditOptions#ApplyTo would make the overall structure easier to see